File: tests/benchmark_performance_tests/benchmark_performance_report.py
import statistics
from typing import List, Tuple, TypedDict


class BenchmarkPerformanceReportConfig:
    model_id: int
    model_name: str
    model_xml_path: str
    dataset_id: int
    dataset_name: int
    dataset_path: str
    device_name: str = 'CPU'
    inference_time: int = 20
    experiments_number: int = 10
    experiments: Tuple[Tuple[int, int]] = ((1, 1), (2, 2), (1, 4))

    def __init__(self, model_xml_path: str, dataset_path: str):
        self.model_xml_path = model_xml_path
        self.dataset_path = dataset_path

    def to_dict(self) -> dict:
        return {
            'model_xml_path': self.model_xml_path,
            'dataset_path': self.dataset_path,
            'device_name': self.device_name,
            'inference_time': self.inference_time,
            'experiments_number': self.experiments_number,
            'experiments': self.experiments,
        }


class BenchmarkInference(TypedDict):
    batch: int
    streams: int
    throughput: float


class CLIExperiment:
    batch: int
    streams: int
    _inferences: List = []
    average_throughput: float = None

    def __init__(self, batch: int, streams: int):
        self.batch = batch
        self.streams = streams

    class _CLIExperimentDict(TypedDict):
        batch: int
        streams: int
        inferences: List[BenchmarkInference]
        average_throughput: float

    @property
    def key(self) -> str:
        return f'b{self.batch};s{self.streams}'

    @property
    def inferences(self) -> List[BenchmarkInference]:
        return self._inferences

    @inferences.setter
    def inferences(self, inferences: List[BenchmarkInference]):
        self._inferences = inferences
        average_throughput = statistics.mean([i['throughput'] for i in inferences])
        self.average_throughput = average_throughput

    def to_dict(self) -> dict:
        return self._CLIExperimentDict(batch=self.batch, streams=self.streams, inferences=self.inferences,
                                       average_throughput=self.average_throughput)


class BenchmarkPerformanceReport:
    _default_ui_experiments_dict = {
        'experiments': {},
        'average_throughput_drop': None
    }

    config: BenchmarkPerformanceReportConfig
    cli: dict = {}
    ui: dict = {
        'local': {
            'single': _default_ui_experiments_dict,
            'group': _default_ui_experiments_dict,
        }
    }

    class _BenchmarkPerformanceReportDict(TypedDict):
        config: dict
        cli: dict
        ui: dict

    def __init__(self, config: BenchmarkPerformanceReportConfig):
        self.config = config

    def to_dict(self) -> dict:
        return self._BenchmarkPerformanceReportDict(config=self.config.to_dict(), cli=self.cli, ui=self.ui)

    def add_cli_experiments(self, experiments: List[CLIExperiment]):
        for experiment in experiments:
            self.cli[experiment.key] = experiment.to_dict()

File: client/e2e/src/utils/benchmark-performance-report/benchmark-performance-report.ts
interface IBenchmarkPerformanceReportConfig {
  model_xml_path: string;
  dataset_path: string;
  device_name: 'CPU';
  inference_time: number;
  experiments_number: number;
  experiments: [number, number][];
}

export interface IBenchmarkInference {
  batch: number;
  streams: number;
  throughput: number;
}

interface ICLIBenchmarkExperiment {
  batch: number;
  streams: number;
  inferences: IBenchmarkInference[];
  average_throughput: number | null;
}

export interface IUIBenchmarkExperiment extends ICLIBenchmarkExperiment {
  average_throughput_drop: number | null;
}

export class UIBenchmarkExperiment implements IUIBenchmarkExperiment {
  batch: number;
  streams: number;
  inferences: IBenchmarkInference[];
  average_throughput: number | null;
  average_throughput_drop: number | null;

  constructor(inferences: IBenchmarkInference[]) {
    if (!inferences?.length) {
      throw Error('Empty inference list is passed to UI Benchmark Experiment constructor');
    }
    const [{ batch, streams }] = inferences;
    this.batch = batch;
    this.streams = streams;
    this.inferences = inferences;
    this.average_throughput = UIBenchmarkExperiment.calculateAverageThroughput(inferences);
    this.average_throughput_drop = null;
  }

  get key(): string {
    return UIBenchmarkExperiment.getExperimentKey(this.batch, this.streams);
  }

  static getExperimentKey(batch: number, streams: number): string {
    return `b${batch};s${streams}`;
  }

  static calculateAverageThroughput(inferences: IBenchmarkInference[]): number {
    return inferences.reduce((acc, { throughput }) => acc + throughput, 0) / inferences.length;
  }

  static calculateAverageThroughputDrop(averageThroughput: number, cliThroughput: number): number {
    const averageDrop = (1 - averageThroughput / cliThroughput) * 100;
    return Number(averageDrop.toFixed(2));
  }

  setAverageThroughputDrop(cliThroughput: number): void {
    this.average_throughput_drop = UIBenchmarkExperiment.calculateAverageThroughputDrop(
      this.average_throughput,
      cliThroughput
    );
  }
}

export interface IUIBenchmarkDetails {
  experiments: {
    [key: string]: IUIBenchmarkExperiment;
  };
  average_throughput_drop: number | null;
}

export interface IUIBenchmark {
  single: IUIBenchmarkDetails;
  group: IUIBenchmarkDetails;
}

export interface IBenchmarkPerformanceReport {
  config: IBenchmarkPerformanceReportConfig;
  cli: {
    [key: string]: ICLIBenchmarkExperiment;
  };
  ui: {
    local: IUIBenchmark;
  };
}
